service/handler.go

```diff
@@ -227,6 +227,13 @@ func (h *HandlerService) processPacket(a agent.Agent, p *packet.Packet) error {
 			return fmt.Errorf("Invalid handshake data. Id=%d", a.GetSession().ID())
 		}
 
+		for _, fun := range a.GetSession().GetHandshakeValidators() {
+			if err := fun(handshakeData); err != nil {
+				a.SetStatus(constants.StatusClosed)
+				return fmt.Errorf("Handshake validation failed: %w. Id=%d", err, a.GetSession().ID())
```

Comment:
Is it possible to return specific error codes to clients if the validation fail? That'd be very useful to handle forced updates and server maintenance, for example.

Reply:
Isn't the err here already the one returned by the handshake validator? And if so it is something you have control over already, right?

Reply:
In this step, clients already received a message about the [handshake](https://github.com/topfreegames/pitaya/blob/feature/handshake-validators/service/handler.go#L216).
Any error that may occur after sending the handshake response, just close the session.

Reply:
Hum, I see. I guess it'd be best to return the handshake response only after validation.

From the client perspective, getting a sudden connection close without reason might be very confusing during development, wouldn't it?

Reply:
Yes! I've changed the process to send more comprehensive messages to clients. This first version should not break anything on the client side, but it won't send any information about the error to clients, just a 400 code (Bad Request). A further (maybe breaking) change can be addressed to provide to client a clearer message about what happened in the server.